Update string marshallers to V2 shape

AaronRobinsonMSFT opened the pull request:
This updates the marshallers and the built-in system and NativeAOT's use of them.

/cc @elinor-fung @jkoritzinsky @jkotas 

<details>
  <summary>Generated code examples</summary>

## UTF-16
```csharp
[System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Interop.LibraryImportGenerator", "42.42.42.42")]
[System.Runtime.CompilerServices.SkipLocalsInitAttribute]
public static partial string Strings(string a1, in string a2, ref string a3, out string a4)
{
    System.Runtime.CompilerServices.Unsafe.SkipInit(out a4);
    ushort* __a2_native = default;
    ushort* __a3_native = default;
    ushort* __a4_native = default;
    string __retVal;
    ushort* __retVal_native = default;
    try
    {
        // Marshal - Convert managed data to native data.
        __a2_native = global::System.Runtime.InteropServices.Marshalling.Utf16StringMarshaller.ConvertToUnmanaged(a2);
        __a3_native = global::System.Runtime.InteropServices.Marshalling.Utf16StringMarshaller.ConvertToUnmanaged(a3);
        // Pin - Pin data in preparation for calling the P/Invoke.
        fixed (void* __a1_native = &global::System.Runtime.InteropServices.Marshalling.Utf16StringMarshaller.GetPinnableReference(a1))
        {
            __retVal_native = __PInvoke((ushort*)__a1_native, &__a2_native, &__a3_native, &__a4_native);
        }

        // Unmarshal - Convert native data to managed data.
        __retVal = global::System.Runtime.InteropServices.Marshalling.Utf16StringMarshaller.ConvertToManaged(__retVal_native);
        a4 = global::System.Runtime.InteropServices.Marshalling.Utf16StringMarshaller.ConvertToManaged(__a4_native);
        a3 = global::System.Runtime.InteropServices.Marshalling.Utf16StringMarshaller.ConvertToManaged(__a3_native);
    }
    finally
    {
        // Cleanup - Perform required cleanup.
        global::System.Runtime.InteropServices.Marshalling.Utf16StringMarshaller.Free(__retVal_native);
        global::System.Runtime.InteropServices.Marshalling.Utf16StringMarshaller.Free(__a2_native);
        global::System.Runtime.InteropServices.Marshalling.Utf16StringMarshaller.Free(__a3_native);
        global::System.Runtime.InteropServices.Marshalling.Utf16StringMarshaller.Free(__a4_native);
    }

    return __retVal;
    // Local P/Invoke
    [System.Runtime.InteropServices.DllImportAttribute("Microsoft.Interop.Tests.NativeExportsNE", EntryPoint = "Strings", ExactSpelling = true)]
    static extern unsafe ushort* __PInvoke(ushort* a1, ushort** a2, ushort** a3, ushort** a4);
}
```

## UTF-8
```csharp
[System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Interop.LibraryImportGenerator", "42.42.42.42")]
[System.Runtime.CompilerServices.SkipLocalsInitAttribute]
public static partial string Strings(string a1, in string a2, ref string a3, out string a4)
{
    System.Runtime.CompilerServices.Unsafe.SkipInit(out a4);
    byte* __a1_native = default;
    byte* __a2_native = default;
    byte* __a3_native = default;
    byte* __a4_native = default;
    string __retVal;
    byte* __retVal_native = default;
    // Setup - Perform required setup.
    global::System.Runtime.InteropServices.Marshalling.Utf8StringMarshaller.ManagedToUnmanagedIn __a1_native__marshaller = new();
    global::System.Runtime.InteropServices.Marshalling.Utf8StringMarshaller.ManagedToUnmanagedIn __a2_native__marshaller = new();
    try
    {
        // Marshal - Convert managed data to native data.
        byte* __a1_native__stackptr = stackalloc byte[global::System.Runtime.InteropServices.Marshalling.Utf8StringMarshaller.ManagedToUnmanagedIn.BufferSize];
        __a1_native__marshaller.FromManaged(a1, new System.Span<byte>(__a1_native__stackptr, global::System.Runtime.InteropServices.Marshalling.Utf8StringMarshaller.ManagedToUnmanagedIn.BufferSize));
        byte* __a2_native__stackptr = stackalloc byte[global::System.Runtime.InteropServices.Marshalling.Utf8StringMarshaller.ManagedToUnmanagedIn.BufferSize];
        __a2_native__marshaller.FromManaged(a2, new System.Span<byte>(__a2_native__stackptr, global::System.Runtime.InteropServices.Marshalling.Utf8StringMarshaller.ManagedToUnmanagedIn.BufferSize));
        __a3_native = global::System.Runtime.InteropServices.Marshalling.Utf8StringMarshaller.ConvertToUnmanaged(a3);
        {
            // PinnedMarshal - Convert managed data to native data that requires the managed data to be pinned.
            __a1_native = __a1_native__marshaller.ToUnmanaged();
            __a2_native = __a2_native__marshaller.ToUnmanaged();
            __retVal_native = __PInvoke(__a1_native, &__a2_native, &__a3_native, &__a4_native);
        }

        // Unmarshal - Convert native data to managed data.
        __retVal = global::System.Runtime.InteropServices.Marshalling.Utf8StringMarshaller.ConvertToManaged(__retVal_native);
        a4 = global::System.Runtime.InteropServices.Marshalling.Utf8StringMarshaller.ConvertToManaged(__a4_native);
        a3 = global::System.Runtime.InteropServices.Marshalling.Utf8StringMarshaller.ConvertToManaged(__a3_native);
    }
    finally
    {
        // Cleanup - Perform required cleanup.
        global::System.Runtime.InteropServices.Marshalling.Utf8StringMarshaller.Free(__retVal_native);
        __a1_native__marshaller.Free();
        __a2_native__marshaller.Free();
        global::System.Runtime.InteropServices.Marshalling.Utf8StringMarshaller.Free(__a3_native);
        global::System.Runtime.InteropServices.Marshalling.Utf8StringMarshaller.Free(__a4_native);
    }

    return __retVal;
    // Local P/Invoke
    [System.Runtime.InteropServices.DllImportAttribute("Microsoft.Interop.Tests.NativeExportsNE", EntryPoint = "Strings", ExactSpelling = true)]
    static extern unsafe byte* __PInvoke(byte* a1, byte** a2, byte** a3, byte** a4);
}
```

## ANSI
```csharp
[System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Interop.LibraryImportGenerator", "42.42.42.42")]
[System.Runtime.CompilerServices.SkipLocalsInitAttribute]
public static partial string Strings(string a1, in string a2, ref string a3, out string a4)
{
    System.Runtime.CompilerServices.Unsafe.SkipInit(out a4);
    byte* __a1_native = default;
    byte* __a2_native = default;
    byte* __a3_native = default;
    byte* __a4_native = default;
    string __retVal;
    byte* __retVal_native = default;
    // Setup - Perform required setup.
    global::System.Runtime.InteropServices.Marshalling.AnsiStringMarshaller.ManagedToUnmanagedIn __a1_native__marshaller = new();
    global::System.Runtime.InteropServices.Marshalling.AnsiStringMarshaller.ManagedToUnmanagedIn __a2_native__marshaller = new();
    try
    {
        // Marshal - Convert managed data to native data.
        byte* __a1_native__stackptr = stackalloc byte[global::System.Runtime.InteropServices.Marshalling.AnsiStringMarshaller.ManagedToUnmanagedIn.BufferSize];
        __a1_native__marshaller.FromManaged(a1, new System.Span<byte>(__a1_native__stackptr, global::System.Runtime.InteropServices.Marshalling.AnsiStringMarshaller.ManagedToUnmanagedIn.BufferSize));
        byte* __a2_native__stackptr = stackalloc byte[global::System.Runtime.InteropServices.Marshalling.AnsiStringMarshaller.ManagedToUnmanagedIn.BufferSize];
        __a2_native__marshaller.FromManaged(a2, new System.Span<byte>(__a2_native__stackptr, global::System.Runtime.InteropServices.Marshalling.AnsiStringMarshaller.ManagedToUnmanagedIn.BufferSize));
        __a3_native = global::System.Runtime.InteropServices.Marshalling.AnsiStringMarshaller.ConvertToUnmanaged(a3);
        {
            // PinnedMarshal - Convert managed data to native data that requires the managed data to be pinned.
            __a1_native = __a1_native__marshaller.ToUnmanaged();
            __a2_native = __a2_native__marshaller.ToUnmanaged();
            __retVal_native = __PInvoke(__a1_native, &__a2_native, &__a3_native, &__a4_native);
        }

        // Unmarshal - Convert native data to managed data.
        __retVal = global::System.Runtime.InteropServices.Marshalling.AnsiStringMarshaller.ConvertToManaged(__retVal_native);
        a4 = global::System.Runtime.InteropServices.Marshalling.AnsiStringMarshaller.ConvertToManaged(__a4_native);
        a3 = global::System.Runtime.InteropServices.Marshalling.AnsiStringMarshaller.ConvertToManaged(__a3_native);
    }
    finally
    {
        // Cleanup - Perform required cleanup.
        global::System.Runtime.InteropServices.Marshalling.AnsiStringMarshaller.Free(__retVal_native);
        __a1_native__marshaller.Free();
        __a2_native__marshaller.Free();
        global::System.Runtime.InteropServices.Marshalling.AnsiStringMarshaller.Free(__a3_native);
        global::System.Runtime.InteropServices.Marshalling.AnsiStringMarshaller.Free(__a4_native);
    }

    return __retVal;
    // Local P/Invoke
    [System.Runtime.InteropServices.DllImportAttribute("Microsoft.Interop.Tests.NativeExportsNE", EntryPoint = "Strings", ExactSpelling = true)]
    static extern unsafe byte* __PInvoke(byte* a1, byte** a2, byte** a3, byte** a4);
}
```

## BSTR
```csharp
[System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Interop.LibraryImportGenerator", "42.42.42.42")]
[System.Runtime.CompilerServices.SkipLocalsInitAttribute]
public static partial string Strings(string a1, in string a2, ref string a3, out string a4)
{
    System.Runtime.CompilerServices.Unsafe.SkipInit(out a4);
    ushort* __a1_native = default;
    ushort* __a2_native = default;
    ushort* __a3_native = default;
    ushort* __a4_native = default;
    string __retVal;
    ushort* __retVal_native = default;
    // Setup - Perform required setup.
    global::System.Runtime.InteropServices.Marshalling.BStrStringMarshaller.ManagedToUnmanagedIn __a1_native__marshaller = new();
    global::System.Runtime.InteropServices.Marshalling.BStrStringMarshaller.ManagedToUnmanagedIn __a2_native__marshaller = new();
    try
    {
        // Marshal - Convert managed data to native data.
        byte* __a1_native__stackptr = stackalloc byte[global::System.Runtime.InteropServices.Marshalling.BStrStringMarshaller.ManagedToUnmanagedIn.BufferSize];
        __a1_native__marshaller.FromManaged(a1, new System.Span<byte>(__a1_native__stackptr, global::System.Runtime.InteropServices.Marshalling.BStrStringMarshaller.ManagedToUnmanagedIn.BufferSize));
        byte* __a2_native__stackptr = stackalloc byte[global::System.Runtime.InteropServices.Marshalling.BStrStringMarshaller.ManagedToUnmanagedIn.BufferSize];
        __a2_native__marshaller.FromManaged(a2, new System.Span<byte>(__a2_native__stackptr, global::System.Runtime.InteropServices.Marshalling.BStrStringMarshaller.ManagedToUnmanagedIn.BufferSize));
        __a3_native = global::System.Runtime.InteropServices.Marshalling.BStrStringMarshaller.ConvertToUnmanaged(a3);
        {
            // PinnedMarshal - Convert managed data to native data that requires the managed data to be pinned.
            __a1_native = __a1_native__marshaller.ToUnmanaged();
            __a2_native = __a2_native__marshaller.ToUnmanaged();
            __retVal_native = __PInvoke(__a1_native, &__a2_native, &__a3_native, &__a4_native);
        }

        // Unmarshal - Convert native data to managed data.
        __retVal = global::System.Runtime.InteropServices.Marshalling.BStrStringMarshaller.ConvertToManaged(__retVal_native);
        a4 = global::System.Runtime.InteropServices.Marshalling.BStrStringMarshaller.ConvertToManaged(__a4_native);
        a3 = global::System.Runtime.InteropServices.Marshalling.BStrStringMarshaller.ConvertToManaged(__a3_native);
    }
    finally
    {
        // Cleanup - Perform required cleanup.
        global::System.Runtime.InteropServices.Marshalling.BStrStringMarshaller.Free(__retVal_native);
        __a1_native__marshaller.Free();
        __a2_native__marshaller.Free();
        global::System.Runtime.InteropServices.Marshalling.BStrStringMarshaller.Free(__a3_native);
        global::System.Runtime.InteropServices.Marshalling.BStrStringMarshaller.Free(__a4_native);
    }

    return __retVal;
    // Local P/Invoke
    [System.Runtime.InteropServices.DllImportAttribute("Microsoft.Interop.Tests.NativeExportsNE", EntryPoint = "Strings", ExactSpelling = true)]
    static extern unsafe ushort* __PInvoke(ushort* a1, ushort** a2, ushort** a3, ushort** a4);
}
```

  </details>

AaronRobinsonMSFT committed:
Update to string marshallers to new V2 shape.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into update_strings

AaronRobinsonMSFT committed:
Fix incorrect BStr casing.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/Utf8StringMarshaller.cs:
```diff
@@ -10,93 +10,112 @@ namespace System.Runtime.InteropServices.Marshalling
+                    int exactByteCount = checked(Encoding.UTF8.GetByteCount(managed) + 1); // + 1 for null terminator
+                    if (exactByteCount > buffer.Length)
+                    {
+                        buffer = new Span<byte>((byte*)Marshal.AllocCoTaskMem(exactByteCount), exactByteCount);
```
This can use `NativeMemory.Alloc` now that the `In` mode is separate from the rest.  `NativeMemory.Alloc` is faster than `Marshal.AllocCoTaskMem`.

jkotas replied:
(same for other marshallers)

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/AnsiStringMarshaller.cs:
```diff
@@ -9,90 +9,109 @@ namespace System.Runtime.InteropServices.Marshalling
+            /// <summary>
+            /// Requested buffer size for optimized marshalling.
+            /// </summary>
+            public static int BufferSize { get; } = 0x100;
```
```suggestion
            public static int BufferSize => 0x100;
```
{ get; } = 0x100; means that the value would be cached in readonly static. We do not need that.

AaronRobinsonMSFT replied:
I assume I should update all uses of that pattern?

jkotas replied:
(same for other marshallers)

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/AnsiStringMarshaller.cs:
```diff
@@ -9,90 +9,109 @@ namespace System.Runtime.InteropServices.Marshalling
+            /// Initialize the marshaller with an unmanaged string.
+            /// </summary>
+            /// <param name="unmanaged">An unmanaged string</param>
+            public void FromUnmanaged(byte* unmanaged)
```
We should not need `FromUnmanaged` and `ToManaged` method on the `In` mode marshaller.

jkotas replied:
(same for other string marshallers)

jkotas commented on src/coreclr/vm/ilmarshalers.cpp:
```diff
@@ -1959,7 +1959,7 @@ void ILCUTF8Marshaler::EmitConvertContentsCLRToNative(ILCodeStream* pslILEmit)
 
     if (m_dwInstance == LOCAL_NUM_UNUSED)
-        m_dwInstance = pslILEmit->NewLocal(LocalDesc(CoreLibBinder::GetClass(CLASS__UTF8STRINGMARSHALLER)));
+        m_dwInstance = pslILEmit->NewLocal(LocalDesc(CoreLibBinder::GetClass(CLASS__UTF8STRINGMARSHALLER_IN)));
```
We need to initialize this only when `bPassByValueInOnly` is true now. The remaining cases are stateless and do not need the instance.

jkotas replied:
(same in the AOT compiler version)

jkotas commented on src/coreclr/vm/clsload.cpp:
```diff
@@ -305,13 +327,32 @@ TypeHandle ClassLoader::LoadTypeByNameThrowing(Assembly *pAssembly,
-        RETURN pAssembly->GetLoader()->LoadTypeHandleThrowIfFailed(&nameHandle, level);
-    else
-        RETURN pAssembly->GetLoader()->LoadTypeHandleThrowing(&nameHandle, level);
+    ClassLoader* classLoader = pAssembly->GetLoader();
```
This method is called from number of places. I do not think we want this change in all those other places. It will change the behavior in subtle ways and it may make it impossible to load types with `+` in the type name. `+` in (non-fully qualified) type name is valid and I can imagine things like obfuscators may take advantage of it.

jkotas commented on src/coreclr/vm/ilmarshalers.cpp:
```diff
@@ -2022,14 +2017,12 @@ void ILCUTF8Marshaler::EmitClearNative(ILCodeStream* pslILEmit)
         _ASSERTE(m_dwInstance != LOCAL_NUM_UNUSED);
```
Nit: We can just check `if (m_dwInstance != LOCAL_NUM_UNUSED)` instead of replicating the full condition for `bPassByValueInOnly`

AaronRobinsonMSFT committed:
Use NativeMemory.Alloc/Free where possible.
Avoid initialization of marshaller when not used.
Remove unneccessary unmanaged->managed APIs for In marshallers.

AaronRobinsonMSFT committed:
Add a ClassLoader::LoadTypeByNameThrowing overload
  that can be used to load nested classes.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/AnsiStringMarshaller.cs:
```diff
@@ -9,90 +9,93 @@ namespace System.Runtime.InteropServices.Marshalling
+        /// <param name="unmanaged">An unmanaged string</param>
+        /// <returns>A managed string</returns>
+        public static string? ConvertToManaged(byte* unmanaged)
+            => Marshal.PtrToStringAnsi((nint)unmanaged);
```
```suggestion
            => Marshal.PtrToStringAnsi((IntPtr)unmanaged);
```
Nit: I think it is better to stick with IntPtr in places where it actually means pointer.

jkotas replied:
(multiple places)

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/AnsiStringMarshaller.cs:
```diff
@@ -9,90 +9,93 @@ namespace System.Runtime.InteropServices.Marshalling
+        /// <param name="managed">A managed string</param>
+        /// <returns>An unmanaged string</returns>
+        public static byte* ConvertToUnmanaged(string? managed)
+            => (byte*)Marshal.StringToCoTaskMemAnsi(managed);
```
`Marshal.StringToCoTaskMemAnsi` is single-pass and allocates memory conservatively (it can allocate 2x - 3x more than necessary). The existing `AnsiStringMarshaller` marshaller implementation and the existing built-in marshallers are two-pass and allocate exact amount of memory.

Not sure whether we care enough about this difference. Just pointing it out. Feel free to "won't fix".

jkotas replied:
(same for Utf8 marshaller)

AaronRobinsonMSFT replied:
Bah. It's a fair criticism and shouldn't be hard to do correctly. I'll make the change.

jkotas reviewed (approved):
LGTM. Thank you!

AaronRobinsonMSFT committed:
Allocate precise amounts for UTF8 and ANSI marshallers.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/AnsiStringMarshaller.cs:
```diff
@@ -9,90 +9,102 @@ namespace System.Runtime.InteropServices.Marshalling
+                return null;
+
+            int exactByteCount = Marshal.GetAnsiStringByteCount(managed); // Includes null terminator
+            Span<byte> buffer = new ((byte*)NativeMemory.Alloc((nuint)exactByteCount), exactByteCount);
```
This has to use CoTaskMemAlloc. It is part of the contract for marshalling strings as `ref` and `out`.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/AnsiStringMarshaller.cs:
```diff
@@ -9,90 +9,102 @@ namespace System.Runtime.InteropServices.Marshalling
+            Span<byte> buffer = new ((byte*)NativeMemory.Alloc((nuint)exactByteCount), exactByteCount);
+
+            Marshal.GetAnsiStringBytes(managed, buffer); // Includes null terminator
+            return (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(buffer));
```
Nit: I think storing the pointer in a local and just returning it here would look better.

AaronRobinsonMSFT committed:
Use CoTaskMemAlloc/Free

AaronRobinsonMSFT committed:
Keep allocation in local.
